python/mirascope/llm/content/image.py

```diff
@@ -57,24 +64,117 @@ def from_url(
         download: bool = False,
     ) -> "Image":
         """Create an `Image` from a URL."""
-        raise NotImplementedError
+        if download:
+            response = httpx.get(url)
+            response.raise_for_status()
+            data = base64.b64encode(response.content).decode("utf-8")
+
+            # Infer mime type from content-type header or URL
+            mime_type: ImageMimeType
+            content_type = response.headers.get("content-type", "")
+            if content_type.startswith("image/"):
+                mime_type_str = content_type.split(";")[0]
+                if mime_type_str in MIME_TYPES:
+                    mime_type = mime_type_str
+                else:
+                    mime_type = "image/jpeg"
+            else:
+                ext = Path(urlparse(url).path).suffix.lower()
+                mime_type_map: dict[str, ImageMimeType] = {
+                    ".png": "image/png",
+                    ".jpg": "image/jpeg",
+                    ".jpeg": "image/jpeg",
+                    ".webp": "image/webp",
+                    ".gif": "image/gif",
+                    ".heic": "image/heic",
+                    ".heif": "image/heif",
+                }
+                mime_type = mime_type_map.get(ext, "image/jpeg")
+
+            return cls(
+                source=Base64ImageSource(
+                    type="base64_image_source",
+                    data=data,
+                    mime_type=mime_type,
+                )
+            )
+        else:
+            return cls(source=URLImageSource(type="url_image_source", url=url))
 
     @classmethod
     def from_file(
         cls,
         file_path: str,
         *,
-        mime_type: ImageMimeType | None,
+        mime_type: ImageMimeType | None = None,
     ) -> "Image":
         """Create an `Image` from a file path."""
-        raise NotImplementedError
+        path = Path(file_path)
+        if not path.exists():
+            raise FileNotFoundError(f"File not found: {file_path}")
+
+        with open(path, "rb") as f:
+            data = base64.b64encode(f.read()).decode("utf-8")
+
+        final_mime_type: ImageMimeType
+        if mime_type is None:
+            ext = path.suffix.lower()
+            mime_type_map: dict[str, ImageMimeType] = {
+                ".png": "image/png",
+                ".jpg": "image/jpeg",
+                ".jpeg": "image/jpeg",
+                ".webp": "image/webp",
+                ".gif": "image/gif",
+                ".heic": "image/heic",
+                ".heif": "image/heif",
+            }
+            final_mime_type = mime_type_map.get(ext, "image/jpeg")
+        else:
+            final_mime_type = mime_type
```

Comment:
I understand that only ImageMimeType is accepted, but is it acceptable to accept all values at the runtime level? Does this mean validation is left to Anthropic?

Reply:
Ah, I see. This is a general-purpose method not limited to Anthropic. In that case, it might be fine.